kv/kvserver: TestReplicaCircuitBreaker_ExemptRequests failed

kv/kvserver.TestReplicaCircuitBreaker_ExemptRequests [failed](https://tanzanite.cluster.engflow.com/invocations/default/bb5dd123-bba5-4fb4-888e-eb726d758ec3?testReportRun=2&testReportShard=30&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [507c68f8abb0c9a3f527c66cb42129c2cb096e7d](https://github.com/cockroachdb/cockroach/commits/507c68f8abb0c9a3f527c66cb42129c2cb096e7d):


```
          1421.994ms   1421.758ms    event:kv/kvserver/replica_send.go:190 [n1,s1,r75/1:/{Table/Max-Max}] replica.Send got error: result is ambiguous: replica unavailable: (n1,s1):1 unable to serve request to r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: lost quorum (down: (n2,s2):2); closed timestamp: 1738575736.643927123,0 (2025-02-03 09:42:16); raft status: {"id":"1","term":6,"vote":"1","commit":31,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":31,"progress":{},"leadtransferee":"0"}: have been waiting 1.50s for slow proposal Put [/Table/Max]
    client_replica_circuit_breaker_test.go:733: result is ambiguous: replica unavailable: (n1,s1):1 unable to serve request to r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: lost quorum (down: (n2,s2):2); closed timestamp: 1738575736.643927123,0 (2025-02-03 09:42:16); raft status: {"id":"1","term":6,"vote":"1","commit":31,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":31,"progress":{},"leadtransferee":"0"}: have been waiting 1.50s for slow proposal Put [/Table/Max]
    client_replica_circuit_breaker_test.go:1411: info:
             0.000ms      0.000ms    === operation:SendCtx(Get) _verbose:1
             0.078ms      0.078ms    event:kv/kvserver/replica_send.go:171 [n1,s1,r75/1:/{Table/Max-Max}] read-only path
             0.126ms      0.048ms    event:kv/kvserver/concurrency/concurrency_manager.go:209 [n1,s1,r75/1:/{Table/Max-Max}] sequencing request
             0.166ms      0.040ms    event:kv/kvserver/concurrency/concurrency_manager.go:290 [n1,s1,r75/1:/{Table/Max-Max}] acquiring latches
             0.255ms      0.089ms    event:kv/kvserver/spanlatch/manager.go:580 [n1,s1,r75/1:/{Table/Max-Max}] waiting to acquire read latch /Table/Max@1738575741.075751292,0 for request Get [/Table/Max], held by write latch /Table/Max@1738575739.652649202,0 for request Put [/Table/Max]
             1.128ms      0.873ms    event:kv/kvserver/replica_send.go:190 [n1,s1,r75/1:/{Table/Max-Max}] replica.Send got error: replica unavailable: (n1,s1):1 unable to serve request to r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: lost quorum (down: (n2,s2):2); closed timestamp: 1738575736.643927123,0 (2025-02-03 09:42:16); raft status: {"id":"1","term":6,"vote":"1","commit":31,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":31,"progress":{},"leadtransferee":"0"}: encountered poisoned latch /Table/Max@1738575739.652649202,0
    client_replica_circuit_breaker_test.go:734: replica unavailable: (n1,s1):1 unable to serve request to r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: lost quorum (down: (n2,s2):2); closed timestamp: 1738575736.643927123,0 (2025-02-03 09:42:16); raft status: {"id":"1","term":6,"vote":"1","commit":31,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":31,"progress":{},"leadtransferee":"0"}: encountered poisoned latch /Table/Max@1738575739.652649202,0
    --- FAIL: TestReplicaCircuitBreaker_ExemptRequests/lease-type=LeaseLeader (5.06s)
=== RUN   TestReplicaCircuitBreaker_ExemptRequests/lease-type=LeaseLeader/with-unavailable-lease/AddSSTable
    client_replica_circuit_breaker_test.go:1411: info:
             0.000ms      0.000ms    === operation:SendCtx(AddSSTable) _verbose:1
            39.115ms     39.115ms    event:kv/kvserver/replica_send.go:175 [n1,s1,r75/1:/{Table/Max-Max}] read-write path
            42.140ms      3.025ms    event:kv/kvserver/replica_send.go:190 [n1,s1,r75/1:/{Table/Max-Max}] replica.Send got error: aborted during Replica.Send: context deadline exceeded
    client_replica_circuit_breaker_test.go:698: aborted during Replica.Send: context deadline exceeded
    client_replica_circuit_breaker_test.go:704: 
        	Error Trace:	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:704
        	            				pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1142
        	Error:      	Received unexpected error:
        	            	aborted during Replica.Send: context deadline exceeded
        	            	(1)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
        	            	  | 	pkg/kv/kvserver/replica_send.go:431
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
        	            	  | 	pkg/kv/kvserver/replica_send.go:177
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send
        	            	  | 	pkg/kv/kvserver/replica_send.go:115
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*circuitBreakerTest).SendCtxTS
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1293
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*circuitBreakerTest).SendCtx
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1256
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaCircuitBreaker_ExemptRequests.func1.9
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:693
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*circuitBreakerTest).Run.func1
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1142
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (2) aborted during Replica.Send
        	            	Wraps: (3) context deadline exceeded
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.withPrefix (3) context.deadlineExceededError
        	Test:       	TestReplicaCircuitBreaker_ExemptRequests/lease-type=LeaseLeader/with-unavailable-lease/AddSSTable
        --- FAIL: TestReplicaCircuitBreaker_ExemptRequests/lease-type=LeaseLeader/with-unavailable-lease/AddSSTable (0.06s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=30</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCircuitBreaker_ExemptRequests.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47128

kv/kvserver.TestReplicaCircuitBreaker_ExemptRequests [failed](https://tanzanite.cluster.engflow.com/invocations/default/7857fa0e-980a-4518-ba6c-93cc23bd57aa?testReportRun=2&testReportShard=30&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [13d8c5191614530b50802f77081f15d2ce2ca705](https://github.com/cockroachdb/cockroach/commits/13d8c5191614530b50802f77081f15d2ce2ca705):


```
             4.809ms      0.306ms                event:kv/kvserver/replica_evaluate.go:511 [n1,s1,r75/1:/{Table/Max-Max}] evaluated RequestLease command header:<key:"\372" > lease:<start:<> replica:<node_id:1 store_id:1 replica_id:1 type:VOTER_FULL > proposed_ts:<wall_time:1738920155756201227 > sequence:1 acquisition_type:Request min_expiration:<wall_time:1738920161756201227 > term:6 > prev_lease:<start..., txn=<nil> : resp=header:<> , err=<nil>
             4.900ms      0.091ms                event:kv/kvserver/replica_proposal.go:1049 [n1,s1,r75/1:/{Table/Max-Max}] need consensus on write batch with op count=1
             4.949ms      0.049ms                event:kv/kvserver/replica_raft.go:103 [n1,s1,r75/1:/{Table/Max-Max}] evaluated request
             5.391ms      0.442ms                event:kv/kvserver/replica_raft.go:154 [n1,s1,r75/1:/{Table/Max-Max}] proposing command to write 0 new keys, 0 new values, 0 new intents, write batch size=87 bytes
             5.552ms      0.160ms                event:kv/kvserver/replica_raft.go:268 [n1,s1,r75/1:/{Table/Max-Max}] acquiring proposal quota (265 bytes)
             5.608ms      0.056ms                event:kv/kvserver/replica_raft.go:463 [n1,s1,r75/1:/{Table/Max-Max}] submitting proposal to proposal buffer
            21.512ms     15.904ms                event:kv/kvserver/replica_proposal_buf.go:566 [n1,s1,r75/1:/{Table/Max-Max}] flushing proposal to Raft
            21.577ms      0.065ms                event:kv/kvserver/rafttrace/rafttrace.go:318 [n1,s1,r75/1:/{Table/Max-Max}] too many active raft traces, skipping
            43.426ms     21.849ms                    === operation:local proposal _verbose:1 node:1 store:1 range:75/1:/{Table/Max-Max} raft:
            44.703ms      1.277ms                    event:kv/kvserver/app_batch.go:116 [n1,s1,r75/1:/{Table/Max-Max},raft] applying command
            46.499ms      1.796ms                    event:kv/kvserver/replica_application_state_machine.go:179 [n1,s1,r75/1:/{Table/Max-Max},raft] LocalResult (reply: (err: <nil>), *kvpb.RequestLeaseResponse, #encountered intents: 0, #acquired locks: 0, #resolved locks: 0#updated txns: 0 #end txns: 0, PopulateBarrierResponse:false GossipFirstRange:false MaybeGossipSystemConfig:false MaybeGossipSystemConfigIfHaveFailure:false MaybeAddToSplitQueue:false MaybeGossipNodeLiveness:<nil> 
            47.058ms      0.559ms                    event:kv/kvserver/replica_proposal.go:629 [n1,s1,r75/1:/{Table/Max-Max},raft] new range lease repl=(n1,s1):1 seq=1 start=0,0 term=6 min-exp=1738920161.756201227,0 pro=1738920155.756201227,0 acq=Request following repl=(n1,s1):1 seq=1 start=0,0 term=6 min-exp=1738920149.750541388,0 pro=1738920143.750541388,0 acq=Request
             0.804ms      0.211ms    event:kv/kvserver/replica_range_lease.go:1379 [n1,s1,r75/1:/{Table/Max-Max}] waiting for acquisition/transfer after status {Lease:<empty> Now:0,0 RequestTime:0,0 State:ERROR ErrInfo: Liveness:liveness(nid:0 epo:0 exp:0,0) MinValidObservedTimestamp:0,0 LeaderSupport:{Term:0 LeadSupportUntil:0,0}}
            50.199ms     49.396ms    event:kv/kvserver/replica_range_lease.go:1437 [n1,s1,r75/1:/{Table/Max-Max}] lease acquisition succeeded: repl=(n1,s1):1 seq=1 start=0,0 term=6 min-exp=1738920161.756201227,0 pro=1738920155.756201227,0 acq=Request
            50.275ms      0.076ms    event:kv/kvserver/replica_range_lease.go:1366 [n1,s1,r75/1:/{Table/Max-Max}] valid lease {Lease:repl=(n1,s1):1 seq=1 start=0,0 term=6 min-exp=1738920161.756201227,0 pro=1738920155.756201227,0 acq=Request Now:1738920155.805873477,0 RequestTime:1738920155.755642987,0 State:VALID ErrInfo: Liveness:liveness(nid:0 epo:0 exp:0,0) MinValidObservedTimestamp:0,0 LeaderSupport:{Term:6 LeadSupportUntil:1738920116.756477077,0}}
            50.657ms      0.382ms    event:kv/kvserver/replica_send.go:190 [n1,s1,r75/1:/{Table/Max-Max}] replica.Send got error: aborted during Replica.Send: context deadline exceeded
    client_replica_circuit_breaker_test.go:698: aborted during Replica.Send: context deadline exceeded
    client_replica_circuit_breaker_test.go:704: 
        	Error Trace:	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:704
        	            				pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1142
        	Error:      	Received unexpected error:
        	            	aborted during Replica.Send: context deadline exceeded
        	            	(1)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
        	            	  | 	pkg/kv/kvserver/replica_send.go:431
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
        	            	  | 	pkg/kv/kvserver/replica_send.go:173
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send
        	            	  | 	pkg/kv/kvserver/replica_send.go:115
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*circuitBreakerTest).SendCtxTS
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1293
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*circuitBreakerTest).SendCtx
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1256
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaCircuitBreaker_ExemptRequests.func1.9
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:693
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*circuitBreakerTest).Run.func1
        	            	  | 	pkg/kv/kvserver/client_replica_circuit_breaker_test.go:1142
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (2) aborted during Replica.Send
        	            	Wraps: (3) context deadline exceeded
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.withPrefix (3) context.deadlineExceededError
        	Test:       	TestReplicaCircuitBreaker_ExemptRequests/lease-type=LeaseLeader/with-unavailable-lease/Export
        --- FAIL: TestReplicaCircuitBreaker_ExemptRequests/lease-type=LeaseLeader/with-unavailable-lease/Export (0.05s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=30</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCircuitBreaker_ExemptRequests.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
+++ b/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
@@ -686,15 +686,25 @@ func TestReplicaCircuitBreaker_ExemptRequests(t *testing.T) {
 						skip.IgnoreLintf(t, "subtest does not apply to %s", m)
 					}
 
-					ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
-					defer cancel()
-					const maxWait = 5 * time.Second
-					tBegin := timeutil.Now()
-					err := tc.SendCtx(ctx, n1, req)
 					// For epoch leases usually logs [NotLeaseHolderError] lease
 					// acquisition canceled because context canceled. While for leader
 					// leases, there should be no error since we propose a leader lease
 					// regardless of whether the LeadSupportUntil is in the future or not.
+					// Therefore, we set a longer timeout for leader leases to avoid it
+					// flaking in race/deadlock builds.
+					var ctx context.Context
+					var cancel context.CancelFunc
+					if leaseType == roachpb.LeaseLeader {
+						ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+					} else {
+						ctx, cancel = context.WithTimeout(context.Background(), 50*time.Millisecond)
+					}
+
+					defer cancel()
+					const maxWait = 5 * time.Second
+					tBegin := timeutil.Now()
+					err := tc.SendCtx(ctx, n1, req)
+
 					t.Log(err)
 					switch leaseType {
 					case roachpb.LeaseEpoch:

